lib/__tests__/custom-params.factory.test.ts

```diff
@@ -0,0 +1,48 @@
+import { Factory, DeepPartial, HookFn } from 'fishery';
+
+class DateTime {
+  private constructor(public date: Date) {}
+
+  toISO(): string {
+    return this.date.toISOString();
+  }
+
+  static fromISO(iso: string): DateTime {
+    return new DateTime(new Date(iso));
+  }
+}
+
+type User = {
+  id: string;
+  createdAt: DateTime;
+};
+
+type UserWithOptionalCreatedAt = DeepPartial<User> & {
+  createdAt?: DateTime;
+};
+
+const userFactory = Factory.define<User, {}, User, UserWithOptionalCreatedAt>(
+  ({ params }) => {
+    const created =
+      params.createdAt ?? DateTime.fromISO(new Date().toISOString());
+    return {
+      id: '3',
+      createdAt: created,
+    };
+  },
+);
+
+describe('factory.build', () => {
```

Comment:
It would be good to add this expectation in both the tests where we verify the type of the value in the created object.

```
expect(user.createdAt).toBeInstanceOf(DateTime);
```

Reply:
Actually, createdAt will always be an instance of DateTime because the function must return `T`.

The current signature was 
```typescript
function (input: { params: DeepPartial<T>, ... }): T;
```

I made the example more explicit regarding why this could cause a problem:

You cannot map directly params to values, as shown in the example.

```typescript
Factory.define<User>(
  // @ts-expect-error
  ({ params }) => {
    const createdAt = DateTime.fromISO(new Date().toISOString());
    // As params.createdAt is DeepPartialObject<DateTime>, updatedAt ends up being DeepPartialObject<DateTime>
    const updatedAt = params.createdAt ?? createdAt;

    return {
      id: '3',
      createdAt,
      updatedAt,
    };
  },
);
```